go/worker/common/p2p/p2p.go

```diff
@@ -315,112 +328,24 @@ func messageIdFn(pmsg *pb.Message) string { // nolint: revive
-	if err != nil {
-		return nil, fmt.Errorf("worker/common/p2p: failed to create connection gater: %w", err)
-	}
-
-	// Block peers specified in the blacklist.
-	blacklist := viper.GetStringSlice(CfgP2PBlockedPeerIPs)
-	for _, blockedIP := range blacklist {
-		parsedIP := net.ParseIP(blockedIP)
-		if parsedIP == nil {
-			return nil, fmt.Errorf("worker/common/p2p: malformed blocked IP: %s", blockedIP)
-		}
-
-		if grr := cg.BlockAddr(parsedIP); grr != nil {
-			return nil, fmt.Errorf("worker/common/p2p: failed to block IP (%s): %w", blockedIP, err)
-		}
-	}
-
-	// Maintain persistent peers.
-	persistentPeers := make(map[core.PeerID]bool)
-	persistentPeersAI := []peer.AddrInfo{}
-	for _, pp := range viper.GetStringSlice(CfgP2PPersistentPeers) {
-		// The persistent peer addresses are in the format pubkey@IP:port,
-		// because we use a similar format elsewhere and it's easier for users
-		// to understand than a multiaddr.
-
-		var addr node.ConsensusAddress
-		if grr := addr.UnmarshalText([]byte(pp)); grr != nil {
-			return nil, fmt.Errorf("worker/common/p2p: malformed persistent peer address (expected pubkey@IP:port): %w", grr)
-		}
-
-		pid, grr := api.PublicKeyToPeerID(addr.ID)
-		if grr != nil {
-			return nil, fmt.Errorf("worker/common/p2p: invalid persistent peer public key (%s): %w", addr.ID, grr)
-		}
-
-		ma, grr := addr.Address.MultiAddress()
-		if grr != nil {
-			return nil, fmt.Errorf("worker/common/p2p: failed to convert persistent peer address to multi address (%s): %w", addr, grr)
-		}
-
-		if _, exists := persistentPeers[pid]; exists {
-			// If we already have this peer ID, append to its addresses.
-			for _, p := range persistentPeersAI {
-				if p.ID == pid {
-					p.Addrs = append(p.Addrs, ma)
-					break
-				}
-			}
-		} else {
-			// Fresh entry.
-			ai := peer.AddrInfo{
-				ID:    pid,
-				Addrs: []multiaddr.Multiaddr{ma},
-			}
-			persistentPeersAI = append(persistentPeersAI, ai)
-		}
-		persistentPeers[pid] = true
-		cm.Protect(pid, "")
+	var cfg PeerConfig
+	if err := cfg.Load(); err != nil {
+		return nil, fmt.Errorf("worker/common/p2p: failed to load peer config: %w", err)
 	}
 
 	// Create the P2P host.
-	host, err := libp2p.New(
-		libp2p.UserAgent(fmt.Sprintf("oasis-core/%s", version.SoftwareVersion)),
-		libp2p.ListenAddrs(sourceMultiAddr),
-		libp2p.Identity(api.SignerToPrivKey(identity.P2PSigner)),
-		libp2p.ConnectionManager(cm),
-		libp2p.ConnectionGater(cg),
-	)
+	cfg.HostConfig.Signer = identity.P2PSigner
+	host, cg, err := NewHost(&cfg.HostConfig)
 	if err != nil {
 		return nil, fmt.Errorf("worker/common/p2p: failed to initialize libp2p host: %w", err)
 	}
 
+	// Initialize bootstrapping.
+	var bootstrapDiscovery *bootstrap.Discovery
+	if cfg.EnableBootstrap {
+		bootstrapDiscovery = bootstrap.NewDiscovery(host, cfg.Seeds)
+	}
+
 	// Initialize the gossipsub router.
 	ctx, ctxCancel := context.WithCancel(context.Background())
 	pubsub, err := pubsub.NewGossipSub(
```

Comment:
Note that there is also a pubsub option [`pubsub.WithDiscovery`](https://pkg.go.dev/github.com/libp2p/go-libp2p-pubsub#WithDiscovery) (see implementation in [`discovery.go`](https://github.com/libp2p/go-libp2p-pubsub/blob/master/discovery.go)) which seems like you could pass it the the discovery implementation and it would handle some of the stuff you implemented here?

Also while looking at what is available in go-libp2p, there is this ["backoff connector"](https://github.com/libp2p/go-libp2p/blob/master/p2p/discovery/backoff/backoffconnector.go) which seems like it could be useful and possibly be shared (there's also [`pubsub.WithDiscoverConnector`](https://github.com/libp2p/go-libp2p-pubsub/blob/v0.8.1/discovery.go#L343)) between both bootstrap discovery and registry-based discovery (currently implemented in the peer manager)? Maybe there are some caveats so you should check first if that is something useful.

Reply:
I'm aware about pubsub discovery, but I prefer to add this in a separate PR for 2 reasons. 
- Every topic in pubsub has its own namespace (network in Tendermint) so without a middleware the seed cannot support more chains with different chain context.
- Rate at which pubsub receives new peers through a channel needs to be throttled (comment inside the code) and maybe also limited (to not disconnect all peers from one topic).

Agree, we should merge peers from all discoveries (bootstrap, PEX, registry) and use the backoff. Thought about doing that after we implement all discoveries (if we are planning to implement PEX also). However, I think the registry peers should have priority as those should be available most of the time.

Libp2p also offers dht discovery using kad-dht routing. We can also use that instead of our own PEX implementation, or both of them. Just not sure how to grade peers from so many sources :smile:

Reply:
Ok, sounds good.

>  However, I think the registry peers should have priority as those should be available most of the time.

Agreed, they should have priority.

> Libp2p also offers dht discovery using kad-dht routing.

Yeah maybe let's not add that just yet.